### PR TITLE
Refactor engine API

### PR DESCRIPTION
**Pull request by aknuds1**
###  What have you changed?

Refactor the engine API.

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

- If there is a document change, please file a PR in ([docs](https://github.com/tikv/tikv/tree/master/docs)) and add the PR number here.
- If this PR should be mentioned in the release note, please update the [release notes](https://github.com/tikv/tikv/blob/master/CHANGELOG.md).

###  Does this PR affect `tidb-ansible`?

No.

###  Refer to a related PR or issue link (optional)

 #4184

### CODE REVIEW

**Comment by siddontang** [comment]
awesome!!! @aknuds1

File: components/engine/src/lib.rs [hunk 1]
```diff
@@ -39,43 +40,50 @@ pub use crate::cf::*;
...
+
+    fn sync_kv(&self) -> Result<()>;
+
+    fn write_raft(&self, wb: &WriteBatch) -> Result<()>;
+
+    fn write_raft_opt(&self, wb: &WriteBatch, opts: &WriteOptions) -> Result<()>;
+
+    fn sync_raft(&self) -> Result<()>;
+
+    fn kv(&self) -> &Arc<DB>;
```
Comment (zhangjinpeng1987) [inline components/engine/src/lib.rs]: So the engine trait depends on RocksDB::DB and WriteBatch, do you have further plan to remove this dependency?

**Comment by overvenus** [comment]
Seems conflict with https://github.com/tikv/tikv/pull/5237 . 🤔

File: components/engine/src/lib.rs (hunk 1 above)
Comment (aknuds1) [reply components/engine/src/lib.rs]: @zhangjinpeng1987 Yep, returning `DB` here made this iteration more manageable.

**Comment by aknuds1** [comment]
> Seems conflict with #5237 . 🤔

@overvenus I wasn't aware of that one. Is that one maybe a better starting point?

**Comment by aknuds1** [comment]
This PR doesn't yet work. I just finished it yesterday so it compiles, but from earlier testing, I noticed I had broken something at launch of the server, probably related to the database location since it complained about inconsistencies.

**Comment by overvenus** [comment]
> I wasn't aware of that one. Is that one maybe a better starting point?

We are sorry for not info you, I do think #5237 is a better starting point.
Basically, #5237 does one thing, abstracting write APIs for engines.
The read APIs is still missing. We would appreciate if you could leave some
comments on #5237 and continue working on the read APIs once #5237 merged.

Cc @5kbpers

**Comment by aknuds1** [comment]
> > I wasn't aware of that one. Is that one maybe a better starting point?
> 
> We are sorry for not info you, I do think #5237 is a better starting point.
> Basically, #5237 does one thing, abstracting write APIs for engines.
> The read APIs is still missing. We would appreciate if you could leave some
> comments on #5237 and continue working on the read APIs once #5237 merged.
> 
> Cc @5kbpers

@overvenus OK, sounds good to me. I'll try reading #5237 when I have the time, and see if I can make any useful comments :) Maybe I can supply some more follow-up PRs after it's merged.

I'll be traveling to Vancouver for a couple of weeks of holidays tomorrow though, so my availability will be limited until the 20th.

**Comment by brson** [comment]
@aknuds1 I'm so happy to see you here! I left some comments directed toward both you and @5kbpers over on their PR: https://github.com/tikv/tikv/pull/5237#issuecomment-527675833

**Pull request closed without merging**